### PR TITLE
EE-8027: Thread profiling data production

### PR DESCRIPTION
**Pull request by IvanAndresFritzler**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ComponentProcessingStrategyReactiveProcessorBuilder.java [hunk 1]
```diff
@@ -137,11 +138,12 @@ public ReactiveProcessor build() {
     Optional<ProfilingDataProducer<ComponentProcessingStrategyProfilingEventContext>> psFlowMessagePassingDataProducer =
         dataProducerFromProfilingService(PS_FLOW_MESSAGE_PASSING);
 
-    // location
+    // Component data
     ComponentLocation location = getLocation(processor);
```
Comment (eze210) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ComponentProcessingStrategyReactiveProcessorBuilder.java:142]: Is something missing here? getComponentData or something like that?

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ReactorPublisherBuilder.java [hunk 2]
```diff
@@ -201,6 +224,22 @@ private FluxBuilder(Flux<CoreEvent> flux) {
...
     }
 
+    @Override
+    public ReactorPublisherBuilder<Flux<CoreEvent>> setTaskContext(ProfilingService profilingService, String artifactId,
+                                                                   String artifactType,
+                                                                   ComponentLocation location) {
+      if (profilingService != null) {
+        flux =
+            flux.doOnNext(coreEvent -> {
+              TracingContext taskTracingContext =
```
Comment (eze210) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ReactorPublisherBuilder.java:234]: Avoid this duplicate code

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/DefaultComponentMetadata.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  @Override
+  public String getArtifactType() {
+    return artifactType;
+  }
+
+  @Override
+  public Optional<ComponentLocation> getComponentLocation() {
+    return Optional.ofNullable(componentLocation);
```
Comment (eze210) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/DefaultComponentMetadata.java:46]: static import

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/DefaultTaskTracingContext.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+  private final ComponentMetadata componentMetadata;
+
+  public DefaultTaskTracingContext(ComponentMetadata componentMetadata) {
+    this.componentMetadata = componentMetadata;
+  }
+
+  @Override
+  public Optional<ComponentMetadata> getRunningComponentMetadata() {
+    return Optional.ofNullable(componentMetadata);
```
Comment (eze210) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/DefaultTaskTracingContext.java:24]: static import

**Review by eze210 — CHANGES_REQUESTED** [review]

**Comment by IvanAndresFritzler** [comment]
--validate

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ComponentProcessingStrategyReactiveProcessorBuilder.java (hunk 1 above)
Comment (IvanAndresFritzler) [reply core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ComponentProcessingStrategyReactiveProcessorBuilder.java:151]: I don't think so.

File: core-tests/src/test/java/org/mule/runtime/core/api/profiling/tracing/ThreadLocalTracingServiceTestCase.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+@Feature(PROFILING)
+@Story(DEFAULT_PROFILING_SERVICE)
+public class ThreadLocalTracingServiceTestCase {
+
+  @Rule
+  public MockitoRule mockitoRule = MockitoJUnit.rule();
+
+  private TracingService threadLocalTracingService;
+  @Mock
```
Comment (fsgonz) [inline core-tests/src/test/java/org/mule/runtime/core/api/profiling/tracing/ThreadLocalTracingServiceTestCase.java:42]: space

File: core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/AbstractProcessingStrategyTestCase.java [hunk 6]
```diff
@@ -674,6 +688,14 @@ protected void assertProcessingStrategyProfiling() {
     }
   }
 
+  protected void assertProcessingStrategyTracing() {
+    if (enableProfilingServiceProperty.getValue().equals("true")) {
```
Comment (fsgonz) [inline core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/AbstractProcessingStrategyTestCase.java:692]: use Boolean constant for the string.

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ComponentProcessingStrategyReactiveProcessorBuilder.java [hunk 7]
```diff
@@ -137,11 +147,16 @@ public ReactiveProcessor build() {
     Optional<ProfilingDataProducer<ComponentProcessingStrategyProfilingEventContext>> psFlowMessagePassingDataProducer =
         dataProducerFromProfilingService(PS_FLOW_MESSAGE_PASSING);
 
-    // location
+    // Component data
```
Comment (fsgonz) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ComponentProcessingStrategyReactiveProcessorBuilder.java:150]: Component data?

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ReactorPublisherBuilder.java [hunk 8]
```diff
@@ -201,6 +220,18 @@ private FluxBuilder(Flux<CoreEvent> flux) {
       return this;
     }
 
+    @Override
+    public ReactorPublisherBuilder<Flux<CoreEvent>> setTracingContext(ProfilingService profilingService, String artifactId,
+                                                                      String artifactType,
+                                                                      ComponentLocation componentLocation) {
+      if (profilingService != null) {
+        flux =
+            flux.doOnNext(coreEvent -> setTracingContext(profilingService, artifactId, artifactType, componentLocation,
```
Comment (fsgonz) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ReactorPublisherBuilder.java:229]: should this be done as part of the chain of the processing strategy? Isn't other way without modifying the chain?

**Review by fsgonz — COMMENTED** [review]
Verif

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ComponentProcessingStrategyReactiveProcessorBuilder.java (hunk 7 above)
Comment (IvanAndresFritzler) [reply core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ComponentProcessingStrategyReactiveProcessorBuilder.java:150]: Leaving as it was before.

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/TransactionAwareStreamEmitterProcessingStrategyDecorator.java [hunk 9]
```diff
@@ -104,14 +106,17 @@ public ReactiveProcessor onPipeline(ReactiveProcessor pipeline) {
...
             // a blocking processing strategy that does not involve a thread switch.
-            return buildFlux(pub)
+            ReactorPublisherBuilder<Flux<CoreEvent>> reactorPublisherBuilder = buildFlux(pub)
                 .profileProcessingStrategyEvent(location, getDataProducer(PS_SCHEDULING_FLOW_EXECUTION), artifactId, artifactType)
                 .profileProcessingStrategyEvent(location, getDataProducer(STARTING_FLOW_EXECUTION), artifactId, artifactType)
                 .transform(BLOCKING_PROCESSING_STRATEGY_INSTANCE.onPipeline(pipeline))
-                .profileProcessingStrategyEvent(location, getDataProducer(FLOW_EXECUTED), artifactId, artifactType)
-                .build();
+                .profileProcessingStrategyEvent(location, getDataProducer(FLOW_EXECUTED), artifactId, artifactType);
+            if (featureFlags.isEnabled(ENABLE_PROFILING_SERVICE)) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/TransactionAwareStreamEmitterProcessingStrategyDecorator.java:116]: are we 100% sure we want to use the feature flag mechanism for this? the FF were originally design as a way to maintain backward compatibility when fixing things.

**Review by elrodro83 — DISMISSED** [review]

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/TransactionAwareStreamEmitterProcessingStrategyDecorator.java (hunk 9 above)
Comment (IvanAndresFritzler) [reply core/src/main/java/org/mule/runtime/core/internal/processor/strategy/TransactionAwareStreamEmitterProcessingStrategyDecorator.java:116]: We currently working in the implementation of a better mechanism for this. The idea is to refactor the places where we ask for this feature flag in order to start using such mechanism as soon as it is ready.

**Comment by IvanAndresFritzler** [comment]
--validate

**Comment by IvanAndresFritzler** [comment]
--validate

File: core/src/main/java/org/mule/runtime/core/internal/profiling/TaskSchedulingDataProducerProvider.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+import org.mule.runtime.api.profiling.ProfilingEventContext;
+import org.mule.runtime.api.profiling.ProfilingProducerScope;
+import org.mule.runtime.api.profiling.type.ProfilingEventType;
+import org.mule.runtime.api.profiling.type.context.TaskSchedulingProfilingEventContext;
+import org.mule.runtime.core.api.event.CoreEvent;
+import org.mule.runtime.core.internal.profiling.producer.TaskSchedulingProfilingDataProducer;
+import org.mule.runtime.feature.internal.config.profiling.ProfilingFeatureFlaggingService;
+
+public class TaskSchedulingDataProducerProvider
```
Comment (fsgonz) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/TaskSchedulingDataProducerProvider.java:17]: add jdoc.

**Review by fsgonz — APPROVED** [review]

**Review by fsgonz — APPROVED** [review]

**Review by fsgonz — DISMISSED** [review]

**Review by fsgonz — APPROVED** [review]

**Pull request merged**